### ISSUE TITLE
OSGI/Equinox: Classloader Deadlock on startup 

### ISSUE THREAD

**Issue by chrisrueger** [brettwooldridge/HikariCP#806]
I am running into a problem when Hikari starts up and getting the following Exception. 
Hikari version: HikariCP-2.4.3.jar

```
 2017-01-09 12:18:17,335 WARN  - While loading class "com.zaxxer.hikari.pool.PoolEntry", thread "Thread[Hikari housekeeper (pool springHikariCP),5,main]" timed out waiting (221936ms) for thread "Thread[Start Level: Equinox Container: a055f8be-5cd6-0016-1481-89c49f98de16,5,main]" to finish starting bundle "myBundle [17]". To avoid deadlock, thread "Thread[Hikari housekeeper (pool springHikariCP),5,main]" is proceeding but "com.zaxxer.hikari.pool.PoolEntry" may not be fully initialized.: org.eclipse.osgi_3.10.102.v20160118-1700 [0] - null - [myBundle.Activator$NLogListener@72d5fd0c] c.a.d.o.o.s.Activator
org.osgi.framework.BundleException: Unable to acquire the state change lock for the module: osgi.identity; osgi.identity="myBundle"; type="osgi.bundle"; version:Version="1.0.0" [id=17] STARTED [STARTED]
	at org.eclipse.osgi.container.Module.lockStateChange(Module.java:329) ~[na:na]
	at org.eclipse.osgi.container.Module.start(Module.java:389) ~[na:na]
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:454) ~[na:na]
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.postFindLocalClass(EclipseLazyStarter.java:107) ~[na:na]
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClass(ClasspathManager.java:531) ~[na:na]
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.findLocalClass(ModuleClassLoader.java:324) ~[na:na]
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalClass(BundleLoader.java:327) ~[na:na]
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:402) ~[na:na]
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:352) ~[na:na]
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:344) ~[na:na]
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:160) ~[na:na]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_111]
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:171) ~[na:na]
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:436) ~[na:na]
	at com.zaxxer.hikari.pool.HikariPool.access$500(HikariPool.java:65) ~[na:na]
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:567) ~[na:na]
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:560) ~[na:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111]

```

I investigated and I think I know why:

- we are using OSGI (Eclipse Equinox)
- in Equinox you have bundles, each bundle has an Activator class where the startup happens (start() method gets called by the Equinox OSGI framework)
- each Bundle has an own Classloader
- within this start() method we start Hikari
- During this startup Hikari uses several ThreadPoolExecutors to start background threads e.g. ThreadPoolExecutor addConnectionExecutor
- these Background Threads start doing things like this:
`
PoolEntry newPoolEntry() throws Exception
   {
      return new PoolEntry(newConnection(), this, isReadOnly, isAutoCommit);
   }
`

- now the problem starts:
- the Background thread is starting and trying to load e.g. **PoolEntry** while the surrounding bundle is still starting. 
- it seems that Equinox has a Classloading Locking-mechanism, so that the classloader cannot load  **PoolEntry**, because the Classloader is still locked (because we are in the middle of the bundle startup)

![image](https://cloud.githubusercontent.com/assets/188422/21764876/a7c5f7e4-d665-11e6-8b80-64e0bed665f5.png)

**Questions:**
I am not sure wether this is an Hikari or Equinox issue. 
Do you have any ideas how to get around this? e.g. is there some way to postpone the ThreadPool starts to later until the Bundle startup (which is initializing Hikari) has been finished?

**Comment by brettwooldridge** [comment]
@chrisrueger Try searching for "classloader" on [this page](http://webcache.googleusercontent.com/search?q=cache:1lo5NXl5YcUJ:help.eclipse.org/kepler/topic/org.eclipse.platform.doc.isv/reference/misc/runtime-options.html+&cd=3&hl=en&ct=clnk&gl=jp&client=safari).  Some possibilities are:
* ``osgi.classloader.singleThreadLoads=true``
* ``osgi.classloader.lock=classname``
* ``osgi.classloader.type=parallel``

We used Equinox for many years (at least 10) as our application platform and this still seems strange to me.  Even with the ClassLoader lock, I would have expected the thread(s) started by HikariCP to block until the thread running the activator completed, at which point the lock would be released and they would continue.  We *do* have an OSGi unit test, but it is based on Apache Felix, so obviously the ClassLoading internals are different and this issue is not encountered.

**Comment by chrisrueger** [comment]
@brettwooldridge I forgot to mention, that the application still starts up, but it takes much much longer. I think the deadlock timeout is 5s and we see 6-7 of such exceptions for each startup, so each startup takes around 30-40s longer. So I tried investigating. 

Thanks for the pointers on the Equinox settings. I will try that.

**Comment by chrisrueger** [comment]
Non of the options above brought an effect so far (some more digging needed). 
The only thing which helped and lessened the pain was VM setting:

`-Dosgi.module.lock.timeout=1 `

This reduces the timeout from default 5 seconds to 1s. Definitely speeds up startup time.  When we don't discover other downsides, we can live with the workaround.

Thanks for your help.

**Comment by brettwooldridge** [comment]
@chrisrueger I just re-read your original report above and I have a question.  Are you embedding the HikariCP library jar in your own bundle?  Maybe you didn't know it, but the HikariCP jar is itself an OSGi Bundle:
```
Manifest-Version: 1.0
Bnd-LastModified: 1484110068310
Build-Jdk: 1.8.0_73
Built-By: brettw
Bundle-Description: Ultimate JDBC Connection Pool
Bundle-DocURL: https://github.com/brettwooldridge
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-ManifestVersion: 2
Bundle-Name: HikariCP
Bundle-SymbolicName: com.zaxxer.HikariCP
Bundle-Vendor: Zaxxer.com
Bundle-Version: 2.5.1
Created-By: Apache Maven Bundle Plugin
DynamicImport-Package: *
Export-Package: com.zaxxer.hikari;version="2.5.1";uses:="com.zaxxer.hi
 kari.metrics,javax.naming,javax.naming.spi,javax.sql",com.zaxxer.hika
 ri.hibernate;version="2.5.1";uses:="com.zaxxer.hikari,org.hibernate,o
 rg.hibernate.engine.jdbc.connections.spi,org.hibernate.service.spi",c
 om.zaxxer.hikari.metrics;version="2.5.1",com.zaxxer.hikari.pool;versi
 on="2.5.1";uses:="com.zaxxer.hikari,com.zaxxer.hikari.metrics,com.zax
 xer.hikari.util,javax.sql",com.zaxxer.hikari.util;version="2.5.1";use
 s:="javax.sql"
Import-Package: javax.management,javax.naming,javax.naming.spi,javax.s
 ql,com.codahale.metrics;resolution:=optional;version="[3.1,4)",com.co
 dahale.metrics.health;resolution:=optional;version="[3.1,4)",org.slf4
 j;version="[1.6,2)",org.hibernate;resolution:=optional;version="[5.2,
 6)",org.hibernate.engine.jdbc.connections.spi;resolution:=optional;ve
 rsion="[5.2,6)",org.hibernate.service;resolution:=optional;version="[
 5.2,6)",org.hibernate.service.spi;resolution:=optional;version="[5.2,
 6)",javax.sql.rowset,javax.sql.rowset.serial,javax.sql.rowset.spi,org
 .hibernate.cfg;resolution:=optional;version="[5.2,6)"
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.2.0.201605172007
```
Have you tried deploying HikariCP as a Bundle, importing the ``HikariConfig`` and ``HikariDataSource`` classes into your Bundle, and starting the pool in your BundleActivator as you currently do?

**Comment by chrisrueger** [comment]
@brettwooldridge Yes, I have embedded the jar in the local bundle (bundle classpath). No I didn't know that HikariCP is OSGI bundle. Usually I check that but here I forgot. Thanks for the info. To me it sounds like it could solve my problem, because then HikariCP is loaded with its own exclusive classloader which should not have any locks. I will try that.

**Comment by brettwooldridge** [comment]
Let me know how it goes.

**Comment by chrisrueger** [comment]
We have solved the issue. 
It was a configuration issue on our end. One of the Start-Levels of our bundles was too low. There was a **bundle B** which had a start-level lower than the **bundle A** from which they imported dependencies from. **Bundle A** is the one contain the embedded HikariCP.jar
So I think what Equinox did here is to start **bundle A** first but somehow locked the Classloader (I have not digged deeper to see what exactly happens). 

As soon as we increased the start-level of **Bundle B** to be higher than **Bundle A**, there are no locks anymore and startup is lightning fast.

Thanks for your input and sorry for bothering.
